---
title: "Enrichment-Analyses-with-clusterProfiler"
author: "Eileen Owens"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dpi = 300,
                      cache = TRUE,
                      warning = FALSE,
                      keep_md = TRUE,
                      fig.width=12, fig.height=8,
                      fig.fullwidth=TRUE,
                      message=FALSE)
```
# Introduction
The purpose of this analysis is to perform GSEA and other enrichment analyses to evaluate the enrichment of differential gene expression data (DESeq2 output) for gene sets in the Molecular Signatures Database (MSigDB) and Gene Ontology terms using the package clusterProfiler.
```{r installation, include=FALSE}
# Load necessary packages
library(BiocManager)
library(DESeq2)
library(clusterProfiler)
library(corrplot)
library(dplyr)
library(gson)
library(pathview)
library(enrichplot)
library(ggplot2)
library(ggnewscale)
library(msigdbr)
library(DOSE)
library(ReactomePA)
library(tidyverse)
library(GSEABase)
library(RColorBrewer)
library(ggridges)
library(readr)
library(knitr)
library(ggupset)
```
# Data input
## Import DESeq2 data
Input: DESeq2 results in .CSV format for various comparisons of interest.
```{r data-input}
# Set working directory:
setwd("/path/to/02_scripts") # set working directory to local scripts directory
# Read in data from deseq2
df1 = read.csv("../03_output/DESeq2res.csv", header=TRUE) # replace with correct path and filename of DESeq2 results csv file
```
## Rank genes
There is no consensus for the what ranking metric should be used for GSEA. Popular options from the DESeq2 output include ranking by the 'log2FoldChange' column or the 'stat' column, which is the Wald test statistic. By default, this script ranks genes by the Wald test statistic. To rank by log2 fold change instead, replace 'stat' with 'log2FoldChange'.
```{r rankgenes}
# Rank genes by stat column of deseq2 result
deg_genes_list1 <- df1$stat # Creates a vector of the stat column from DESeq2 output
names(deg_genes_list1) <- df1$gene_name # Names vector with gene symbols
deg_genes_list1 <- na.omit(deg_genes_list1) # Omits any NA values
deg_genes_list1 = sort(deg_genes_list1, decreasing = TRUE) # Sorts in decreasing order
deg_names1 <- names(deg_genes_list1) # Makes a vector of just the ranked gene names
```
Retrieve the collections of interest from MSigDB:
```{r msigdb-import}
# H = hallmark gene sets, C1 = positional gene sets, C2 = curated gene sets, C3 = motif gene sets, C4 = computational gene sets, C5 = GO gene sets, C6 = oncogenic signatures, C7 = immunologic signatures.
m_t2g_C5 <- msigdbr(species = "Homo sapiens", category = "C5") %>%
  dplyr::select(gs_name, gene_symbol)

m_t2g_H <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, gene_symbol)

m_t2g_all <- msigdbr(species = "Homo sapiens") %>%
  dplyr::select(gs_name, gene_symbol)
```

# Gene set enrichment analysis
## Using all gene sets in the C5 (gene ontology) MSigDB category
### Enrichment analysis
```{r gsea_C5}
gse1_msigdb_C5 <- GSEA(deg_genes_list1, # Use the sorted (ranked) list of gene symbols *and* stat ranking column
                  exponent = 1,
                  nPerm = 10000,
                  minGSSize = 1,
                  maxGSSize = 800,
                  pvalueCutoff = 0.05,
                  pAdjustMethod = "BH",
                  TERM2GENE = m_t2g_C5,
                  verbose = TRUE,
                  by = "fgsea")
```

### Dotplot
```{r dotplotC5}
gse1_msigdb_C5 %>%
  dotplot(showCategory = 15, x = "NES", split=".sign") +
  scale_colour_viridis_c(name = "Adjusted\nP-value",
                         option="H") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  geom_vline(xintercept = 0, linetype=2) +
  labs(title="GSEA, MSigDB C5 Gene Sets", y = "Gene Set") +
  theme(plot.title = element_text(hjust = 0.5))
```

### Ridgeline plot
```{r ridgeplotC5, warning=FALSE}
ridgeplot(gse1_msigdb_C5, showCategory=15) + labs(x = "enrichment distribution", title="GSEA, MSigDB C5 Gene Sets", y = "Gene Set") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  theme(plot.title = element_text(hjust = 0.5))
```

## Using all gene sets in the H (hallmark) MSigDB category

### Enrichment analysis
```{r gsea_H}
gse1_msigdb_H <- GSEA(deg_genes_list1, # Use the sorted (ranked) list of gene symbols *and* stat ranking column
                  exponent = 1,
                  nPerm = 10000,
                  minGSSize = 1,
                  maxGSSize = 800,
                  pvalueCutoff = 0.05,
                  pAdjustMethod = "BH",
                  TERM2GENE = m_t2g_H,
                  verbose = TRUE,
                  by = "fgsea")
```

### Dotplot
```{r dotplotH}
gse1_msigdb_H %>%
  dotplot(showCategory = 15, x = "NES", split=".sign") +
  scale_colour_viridis_c(name = "Adjusted\nP-value",
                         option="H") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  geom_vline(xintercept = 0, linetype=2) +
  labs(title="GSEA, MSigDB Hallmark Gene Sets", y = "Gene Set") +
  theme(plot.title = element_text(hjust = 0.5))

```

### Ridgeline plot
```{r ridgeplotH, message=FALSE, warning=FALSE}

ridgeplot(gse1_msigdb_H, showCategory=15) + labs(x = "enrichment distribution", title="GSEA, MSigDB Hallmark Gene Sets", y = "Gene Set") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  theme(plot.title = element_text(hjust = 0.5))

```

## Using all gene sets from all MSigDB categories
```{r gsea_all}

gse1_msigdb_all <- GSEA(deg_genes_list1, # Use the sorted (ranked) list of gene symbols *and* stat ranking column
                  exponent = 1,
                  nPerm = 10000,
                  minGSSize = 1,
                  maxGSSize = 800,
                  pvalueCutoff = 0.05,
                  pAdjustMethod = "BH",
                  TERM2GENE = m_t2g_all,
                  verbose = TRUE,
                  by = "fgsea")

```
### Dotplot
```{r dotplotAll}

gse1_msigdb_all %>%
  dotplot(showCategory = 15, x = "NES", split=".sign") +
  scale_colour_viridis_c(name = "Adjusted\nP-value",
                         option="H") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  geom_vline(xintercept = 0, linetype=2) +
  labs(title="GSEA, All MSigDB Gene Sets", y = "Gene Set") +
  theme(plot.title = element_text(hjust = 0.5))

```

### Ridgeline plot
```{r ridgeplotAll, warning=FALSE}

ridgeplot(gse1_msigdb_all, showCategory=15) + labs(x = "enrichment distribution", title="GSEA, All MSigDB Gene Sets", y = "Gene Set") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  theme(plot.title = element_text(hjust = 0.5))

```

### Export results
Export GSEA results for all significant results in all MSigDB categories as CSV files.
```{r data-export1}
gse1_msigdb_df <- as.data.frame(gse1_msigdb_all)
write.csv(gse1_msigdb_df, file="../03_output/YYMMDD_gsea_HallmarkMSigDBGeneSets.csv") # Replace with desired output path and filename
```

## Using specific gene sets of interest from MSigDB

### Upload gmt files
Input: Gene sets in .gmt format.
```{r one-gmt-import, warning=FALSE}
# If you are importing one gmt file, run this code chunk and exclude the following code chunk with eval=FALSE, include=FALSE.
gmt <- read.gmt("../01_input/filename.gmt") # provide path to gmt file
```

```{r mutliple-gmt-import, warning=FALSE}
# If you are importing multiple GMT files, run this code chunk and exclude the preceding code chunk with eval=FALSE, include=FALSE.

setwd("../01_input/GeneSets") # set working directory to where GMT files are located
gmtFiles <- list.files(pattern = "\\.gmt", full.names = TRUE) # Get list of all .gmt files in this directory
gmtTables <- lapply(gmtFiles, read.gmt) # Apply the read.gmt function to the list of .gmt files and save as a variable
gmt <- do.call(rbind, gmtTables) # Rbind files

setwd("path/to/02_scripts") # Set working directory back to default for the session.
```

### Enrichment analysis
```{r gsea_importedgmt, warning=FALSE}
## Perform GSEA with order ranked gene list and TERM2GENE = the merged data frames of our annotated gene sets of interest.
gse1 <- GSEA(deg_genes_list1, # Use the sorted list of gene symbols *and* stat ranking column
                   exponent = 1,
                   pvalueCutoff = 0.05,
                   pAdjustMethod = "BH",
                   TERM2GENE = gmt,
                   verbose = TRUE,
                   by = "fgsea")
```

### Export results
```{r data-export2}
gse_df1 <- as.data.frame(gse1)
write.csv(gse_df1, file="../03_output/YYMMDD_gsea_geneSetsOfInterest.csv") # Replace with desired output path and filename
```

### Dotplots
```{r dotplot, warning=FALSE}
gse1 %>%
  dotplot(showCategory = 30, x = "NES") +
  scale_colour_viridis_c(name = "Adjusted\nP-value",
                         option="H") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  geom_vline(xintercept = 0, linetype=2) +
  labs(title="GSEA for <Description of Gene Sets>", y = "Gene Set") +
  theme(plot.title = element_text(hjust = 0.5))
```

### Ridgeline plots
```{r ridgeplot, message=FALSE, warning=FALSE}
ridgeplot(gse1, showCategory=20) + labs(x = "enrichment distribution", title="Enrichment for <Description of Gene Sets>", y = "Gene Set") +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) + # replaces underscores with spaces to allow wrapping of long gene set names
  theme(plot.title = element_text(hjust = 0.5))

```

# Over-representation analysis
## Prepare input
Extract only significant results (padj < 0.05) and those with a log2 fold change >1.
```{r prepare-input}
# Extract significant results
sig_genes_df = subset(df1, padj < 0.05)

# Filter out genes that do not have an associated annotated gene name
sig_genes_df <- sig_genes_df[!grepl("ENSCAF", sig_genes_df$gene_name),] # Keep all rows of the sig_genes data frame whose gene_name column *doesn't* contain the string "ENSCAF"

# Filter on log2fc
sig_genes = sig_genes_df$log2FoldChange

# Name the vector
names(sig_genes) <- sig_genes_df$gene_name

# Omit NA values
sig_genes <- na.omit(sig_genes)

# filter on minimum log2 fold change (log2FoldChange > 1)
sig_up_genes <- names(sig_genes)[sig_genes > 1]
sig_up_genes <- names(sig_genes)[sig_genes > 1]
```

## Perform enrichGO analysis
enrichGO performs an over-representation analysis to determine whether known biological functions or processes are over-represented. https://yulab-smu.top/biomedical-knowledge-mining-book/enrichment-overview.html#ora-algorithm 
```{r enrichGO}
go_enrich <- enrichGO(gene = sig_up_genes,
                           ont = "ALL",
                           universe = names(deg_genes_list1),
                           OrgDb = 'org.Hs.eg.db',
                           keyType = 'SYMBOL',
                           readable = FALSE, # set readable to FALSE since we are already using "Symbol" as the keytype
                           pAdjustMethod = "BH",
                           pvalueCutoff = 0.05,
                           qvalueCutoff = 0.1)
```

Print significant results:
```{r}
kable(go_enrich)
```

## Barplot
```{r barplot}
barplot(go_enrich,
        drop = TRUE,
        showCategory = 15,
        title = "Enriched GO Biological Pathways",
        font.size = 8) +
  scale_y_discrete(labels = ~ str_wrap(gsub('_', ' ', .x), 40)) # replaces underscores with spaces to allow wrapping of long gene set names
```

## Connectivity plot
### All terms
```{r}
# Standard connectivity plot
cnetplot(go_enrich,
         showCategory = 10,
         foldChange = sig_genes,
         categorySize = "geneNum") # "geneNum" sizes nodes by number of genes in that pathway; "pvalue" can be used alternatively to size nodes by enrichment p-value

# Circular connectivity plot
cnetplot(go_enrich,
         showCategory = 10,
         foldChange = sig_genes,
         circular = TRUE,
         colorEdge = TRUE
         )
```

### Terms of interest
```{r cnetsetup, results='hide'}
# View significant results of enrichGO
go_enrich$Description

# Create a variable of just terms of interest identified in the results above
selected_pathways <- c("term 1", "term 2", "term 3", "etc")

```

```{r cnetplot, warning=FALSE}
# Standard connectivity plot
cnetplot(go_enrich,
         showCategory = selected_pathways,
         foldChange = sig_genes,
         categorySize = "geneNum") # "geneNum" sizes nodes by number of genes in that pathway; "pvalue" can be used alternatively to size nodes by enrichment p-value

# Circular connectivity plot
cnetplot(go_enrich,
         showCategory = selected_pathways,
         foldChange = sig_genes,
         circular = TRUE,
         colorEdge = TRUE
         )
```



# Citations
```{r}
sessionInfo()
citation()
```
